File: src/main/java/br/com/blz/testjava/controller/ProductController.java
package br.com.blz.testjava.controller;


import br.com.blz.testjava.model.Product;
import br.com.blz.testjava.service.ProductService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/products/test/")
public class ProductController {

    private final ProductService productService;

    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    @PostMapping
    public ResponseEntity<Void> createProduct(@RequestBody Product product) {
        productService.createProduct(product);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @PutMapping("/{sku}")
    public ResponseEntity<Void> updateProduct(@PathVariable Integer sku, @RequestBody Product updatedProduct) {
        productService.updateProduct(sku, updatedProduct);
        return ResponseEntity.status(HttpStatus.OK).build();
    }

    @GetMapping("/{sku}")
    public ResponseEntity<Product> getProductBySku(@PathVariable Integer sku) {
        Product product = productService.getProductBySku(sku);
        if (product != null) {
            return ResponseEntity.ok(product);
        } else {
            return ResponseEntity.notFound().build();
        }
    }

    @DeleteMapping("/{sku}")
    public ResponseEntity<Void> deleteProductBySku(@PathVariable Integer sku) {
        productService.deleteProductBySku(sku);
        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
    }
}

File: src/main/java/br/com/blz/testjava/controller/ProductController.kt
package br.com.blz.testjava.controller

import br.com.blz.testjava.controller.request.ProductRequest
import br.com.blz.testjava.controller.request.ProductUpdateRequest
import br.com.blz.testjava.model.Product
import br.com.blz.testjava.service.ProductService
import br.com.blz.testjava.toModel
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import javax.validation.Valid


@RestController
@RequestMapping("products")
class ProductController(private val productService: ProductService) {

  @GetMapping
  @ResponseStatus(HttpStatus.OK)
  fun index() = productService.findAll()

  @GetMapping("/{sku}")
  @ResponseStatus(HttpStatus.OK)
  fun search(@PathVariable sku: Long) = productService.findBySku(sku)

  @PostMapping
  @ResponseStatus(HttpStatus.CREATED)
  fun save(@RequestBody @Valid request: ProductRequest) = productService.save(request.toModel())

  @PutMapping("/{sku}")
  @ResponseStatus(HttpStatus.OK)
  fun update(@RequestBody @Valid request: ProductUpdateRequest,
             @PathVariable sku: Long): Product? {

    val product = productService.findBySku(sku)

    return productService.save(request.toModel(product))
  }

  @DeleteMapping("/{sku}")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  fun delete(@PathVariable sku: Long) = productService.delete(sku)
}
